File: scripts/integrate-uthmani.py
import os
import re

SQL_FILE = r'd:\f14\quran-uthmani.sql'
DATA_DIR = r'd:\f14\src\data\quran\verses'

def update_quran():
    print("Reading SQL file...")
    with open(SQL_FILE, 'r', encoding='utf-8') as f:
        sql_content = f.read()

    # Regex to extract: (index, sura, aya, 'text')
    # Example: (66, 1, 1, 'بِسْمِ ٱللَّهِ ٱلرَّحْمَـٰنِ ٱلرَّحِيمِ')
    pattern = re.compile(r"\((\d+),\s*(\d+),\s*(\d+),\s*'(.*?)'\)")
    
    quran_data = {}
    matches = pattern.findall(sql_content)
    print(f"Parsed {len(matches)} ayahs.")

    for idx, sura, aya, text in matches:
        s_num = int(sura)
        a_num = int(aya)
        if s_num not in quran_data:
            quran_data[s_num] = []
        quran_data[s_num].append((a_num, text))

    for s_num in range(1, 115):
        file_path = os.path.join(DATA_DIR, f"surah-{s_num}.ts")
        if not os.path.exists(file_path):
            print(f"Skipping missing file: {file_path}")
            continue

        print(f"Updating Surah {s_num}...")
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        # Preserve existing translations
        translations = {}
        # Simple extraction of translations if they exist
        # number: 1, ... urduTranslation: '...', englishTranslation: '...'
        ayah_blocks = re.findall(r"\{\s*number:\s*(\d+).*?\}", content, re.DOTALL)
        for block_num in ayah_blocks:
            a_num = int(block_num)
            # Find the block text for this ayah
            block_match = re.search(fr"\{\s*number:\s*{a_num}.*?\}", content, re.DOTALL)
            if block_match:
                block_text = block_match.group(0)
                urdu = re.search(r"urduTranslation:\s*['\"](.*?)['\"]", block_text)
                english = re.search(r"englishTranslation:\s*['\"](.*?)['\"]", block_text)
                translations[a_num] = {
                    'urdu': urdu.group(1) if urdu else "",
                    'english': english.group(1) if english else ""
                }

        # Extract metadata
        var_name_match = re.search(r"export const (\w+)", content)
        var_name = var_name_match.group(1) if var_name_match else f"surah{s_num}"
        bismillah = "true" if 'bismillah: true' in content else "false"

        # Construct new ayahs array
        new_ayahs = []
        if s_num in quran_data:
            for a_num, text in sorted(quran_data[s_num]):
                trans = translations.get(a_num, {'urdu': "", 'english': ""})
                # Escape single quotes in text
                safe_text = text.replace("'", "\\'")
                safe_urdu = trans['urdu'].replace("'", "\\'")
                safe_eng = trans['english'].replace("'", "\\'")
                
                ayah_obj = f"""    {{
      number: {a_num},
      arabicText: '{safe_text}',
      urduTranslation: '{safe_urdu}',
      englishTranslation: '{safe_eng}',
    }},"""
                new_ayahs.append(ayah_obj)

        new_content = f"""import {{ SurahVerses }} from './surah-1';

export const {var_name}: SurahVerses = {{
  surahNumber: {s_num},
  bismillah: {bismillah},
  ayahs: [
{"\n".join(new_ayahs)}
  ],
}};
"""
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)

    print("Success!")

if __name__ == "__main__":
    update_quran()

File: scripts/integrate-uthmani.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SQL_FILE = 'd:\\f14\\quran-uthmani.sql';
const DATA_DIR = 'd:\\f14\\src\\data\\quran\\verses';

function updateQuran() {
    console.log("Reading SQL file...");
    const sqlContent = fs.readFileSync(SQL_FILE, 'utf8');

    // Regex to extract: (index, sura, aya, 'text')
    // Values part: (66, 1, 1, '...')
    const pattern = /\((\d+),\s*(\d+),\s*(\d+),\s*'(.*?)'\)/g;
    
    const quranData = {};
    let match;
    let count = 0;
    while ((match = pattern.exec(sqlContent)) !== null) {
        const sNum = parseInt(match[2]);
        const aNum = parseInt(match[3]);
        const text = match[4];
        if (!quranData[sNum]) quranData[sNum] = [];
        quranData[sNum].push({ number: aNum, text });
        count++;
    }
    console.log(`Parsed ${count} ayahs across ${Object.keys(quranData).length} surahs.`);

    for (let sNum = 1; sNum <= 114; sNum++) {
        const filePath = path.join(DATA_DIR, `surah-${sNum}.ts`);
        if (!fs.existsSync(filePath)) {
            console.warn(`Skipping missing file: ${filePath}`);
            continue;
        }

        console.log(`Updating Surah ${sNum}...`);
        const content = fs.readFileSync(filePath, 'utf8');

        // Extract existing translations
        const translations = {};
        const ayahRegex = /\{\s*number:\s*(\d+),[\s\S]*?\}/g;
        let ayahMatch;
        while ((ayahMatch = ayahRegex.exec(content)) !== null) {
            const block = ayahMatch[0];
            const aNum = parseInt(ayahMatch[1]);
            
            // Safer extraction using non-greedy match that respects escaped quotes
            // This covers: property: 'text' or property: "text"
            const urduMatch = block.match(/urduTranslation:\s*(['"])([\s\S]*?)(?<!\\)\1/);
            const engMatch = block.match(/englishTranslation:\s*(['"])([\s\S]*?)(?<!\\)\1/);
            
            translations[aNum] = {
                urdu: urduMatch ? urduMatch[2].replace(/\\'/g, "'").replace(/\\"/g, '"') : "",
                english: engMatch ? engMatch[2].replace(/\\'/g, "'").replace(/\\"/g, '"') : ""
            };
        }

        // Extract metadata
        const varMatch = content.match(/export const (\w+)/);
        const varName = varMatch ? varMatch[1] : `surah${sNum}`;
        const bismillah = content.includes('bismillah: true');

        // Construct new ayahs array
        const sortedAyahs = quranData[sNum] ? quranData[sNum].sort((a, b) => a.number - b.number) : [];
        const bismillahPrefixes = [
            "بِسْمِ ٱللَّهِ ٱلرَّحْمَـٰنِ ٱلرَّحِيمِ ",
            "بِّسْمِ ٱللَّهِ ٱلرَّحْمَـٰنِ ٱلرَّحِيمِ "
        ];

        const newAyahs = sortedAyahs.map(a => {
            const trans = translations[a.number] || { urdu: "", english: "" };
            let arabicText = a.text;

            // Strip Bismillah prefix from Ayah 1 if it's not Surah 1 (Al-Fatiha)
            if (sNum > 1 && a.number === 1) {
                for (const prefix of bismillahPrefixes) {
                    if (arabicText.startsWith(prefix)) {
                        console.log(`Stripping Bismillah from Surah ${sNum} Ayah 1`);
                        arabicText = arabicText.substring(prefix.length);
                        break;
                    }
                }
            }

            // Use JSON.stringify to safely handle all escape characters
            const safeText = JSON.stringify(arabicText);
            const safeUrdu = JSON.stringify(trans.urdu || "");
            const safeEng = JSON.stringify(trans.english || "");
            
            return `    {
      number: ${a.number},
      arabicText: ${safeText},
      urduTranslation: ${safeUrdu},
      englishTranslation: ${safeEng},
    },`;
        }).join('\n');




        const newContent = `import { SurahVerses } from './surah-1';

export const ${varName}: SurahVerses = {
  surahNumber: ${sNum},
  bismillah: ${bismillah},
  ayahs: [
${newAyahs}
  ],
};
`;
        fs.writeFileSync(filePath, newContent, 'utf8');
    }

    console.log("Success!");
}

updateQuran();
